### ISSUE TITLE
Contrastive Loss

### ISSUE THREAD

**Issue by wangyu0627** [HKUDS/EasyRec#3]
cos_sim = self.sim(user_pooler_output.unsqueeze(1),pos_item_pooler_output.unsqueeze(0))
neg_sim= self.sim(user_pooler_output.unsqueeze(1),neg_item_pooler_output.unsqueeze(0))
cos_sim = torch.cat([cos_sim, neg_sim], 1)
        
labels = torch.arange(cos_sim.size(0)).long().to(self.device)
loss_fct = nn.CrossEntropyLoss()
loss = loss_fct(cos_sim, labels)
![image](https://github.com/user-attachments/assets/9c148778-ba30-4dc1-89cb-da375e1b19bf)
Do concating operations cause the numerator to have scores of positive and negative samples? Can you explain how the paper and code correspond or give the source of the explanation?

**Comment by Re-bin** [comment]
Hi 👋

Thank you for your interest in EasyRec! There is an unintentional typo in the formula (the negative sign in the denominator). Then, the entire contrastive learning formula is essentially [cross-entropy loss](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html), where the code calculates the predicted value using cosine similarity and provides positive/negative labels for learning.

Specifically, let's assume the dimension of `user_pooler_output` is `[n, dim]`. Then, the dimension of `pos_item_pooler_output` is also `[n, dim]`, while the dimension of `neg_item_pooler_output` is `[m, dim]`. The cosine similarity calculation can be expressed as:
```python
cos_sim = self.sim(user_pooler_output.unsqueeze(1), pos_item_pooler_output.unsqueeze(0))
```

This will first compute a matrix of size `[n, n]`, where the diagonal represents the values corresponding to positive labels, and the off-diagonal elements correspond to negative labels. Here, we treat other users' positive items as negative items for each user. Next, we calculate the similarity with the user's own negative items:
```python
neg_sim = self.sim(user_pooler_output.unsqueeze(1), neg_item_pooler_output.unsqueeze(0))
```

Finally, we concatenate these two results to obtain a matrix of size `[n, n + m]`, where the positions of positive labels in each row are [0, 1, 2, 3, 4, ...]. I hope the above answer is helpful to you :)

Best regards,
Xubin

**Comment by wangyu0627** [comment]
Why use only diagonal elements to represent values corresponding to positive labels?

**Comment by Re-bin** [comment]
Hi 👋

This is because both `user_pooler_output` and `pos_item_pooler_output` are matrices of shape `[n, dim]`, where the user in the `i`-th row corresponds to the pos_item in the `i`-th row. Therefore, after calculating the similarity matrix of shape `[n, n]` through matrix multiplication with cosine similarity (`[n, dim] x [dim, n]`), the diagonal elements correspond exactly to the similarity values between the respective users and items (i.e., positive user-item pairs).

Best regards,
Xubin

**Comment by wangyu0627** [comment]
Okay, I get it. Thanks for the answer.
